### PR TITLE
Lots of NN Code Goodies

### PR DESCRIPTION
**Pull request by JHiltbrand**
**New Additions:**

- Added the `CustomLayer.py` that is a simple wrapper to any layer and introduces a trainable parameter that could be accessed during training and within a loss function. One would just need to pass the model object to the loss function of interest.
- Added the `CustomOptimizer.py` which specifically wraps the `Adam` (v2) optimizer and allows for specifying different learning rates for different layers. There is necessary synergy between the names of layers and the names of learning rate hyper parameters as specified in the main config dictionary. Each time the `_resource_apply_sparse` method gets called, the name of the layer that is passed is checked and used to pass along the correct learning rate back into tensorflow. Likewise, Adam's changing of the learning rate over time is decoupled between the different components by overriding a few more methods from the `optimizer_v2` class.
- Added the `DataLoader.py` which is a fundamental change in how the input data is manipulated and setup for training. This allows per-batch balancing between signal and background events, Njets categories (both separately within signal and background), and stop mass categories. A change in what an "epoch" is also gets incurred. Events are allowed to appear more than once during an epoch and some may not appear, just due to random choice. The length of an "epoch" stays the same in terms of number of batches, just the batches are not guaranteed to exhaust all events. The default is to use the loader rather than the getter. The loading of events from ROOT files is optimized. The number of files is minimized, only the needed branches are loaded, and pandas has been deprecated in favor of just numpy.

**General Updates:**

- In `Models.py` only the `main_model` is left. The naming convention of layers follows a better pattern and necessary for proper use of the `CustomOptimizer`. Generally, layers and corresponding hyperparameters use the same name "stub". The gradient reversal component of the model has been removed.
- In `train.py` changes have been made to accommodate the use of the `DataLoader` class and functionality. The loss function section has been cleaned up with only the relevant, used loss function remaining. Their names have been adjusted to match the naming of the corresponding hyperparameters. All use of gradient reversal has been removed and making the model has been updated. The calculation of ABCD closure has been added to the disco loss function and controlled with an additional hyperparameter.
- In `Validation.py` changes are again made to reflect the use of the `DataLoader` and the change in names of various items. Also, some minor cleanup and improvement to some plotting functions. A signal type enum is put into the ntuples/dataframes/arrays and is used here to pick the relevant events for making plots. See the `DataLoader` class where this is defined. General cleanup has taken place to avoid confusion about which samples are used to make which plots.

### CODE REVIEW

**Comment by JHiltbrand** [comment]
I dedicate commit c9b2937 to @cros0400. Please review when you can.

File: train.py [hunk 1]
```diff
@@ -196,168 +205,67 @@ def __init__(self, USER, debug, seed, replay, saveAndPrint, hyperconfig, doQuick
...
-
-            val_1 = tf.reshape(y_pred[:,  :1], [-1])
-            val_2 = tf.reshape(y_pred[:, 2:3], [-1])
-            normedweight = tf.ones_like(val_1)
-
-            return val_1_disco_loss + val_2_disco_loss + c * cor.distance_corr(val_1, val_2, normedweight, 1)
-        return discoLossAlt
+            d1 = g.uniform(shape=(), minval=0.0, maxval=1.0)
+            d2 = g.uniform(shape=(), minval=0.0, maxval=1.0)
+
```
Comment (cros0400) [inline train.py:236]: Clean up or add print debug function

File: CustomOptimizer.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+class CustomAdam(Adam):
+
+    # This wrapper to the adam, optimizer_v2 classes keeps track of different learning rates
+    # and decay parameters for specified parts of a neural network model
+    # Thus, an lr, beta_1, and beta_2 are provided for the mass regression component
+    # and then for disc components. Default values are also provided for all other components
+    def __init__(self, mass_reg_lr=1.0, disc_lr=0.001, default_lr=0.001,
+                       mass_reg_beta_1=0.9, mass_reg_beta_2=0.999, disc_beta_1=0.9, disc_beta_2=0.999, 
+                       default_beta_1=0.9, default_beta_2=0.999, epsilon=1e-8, amsgrad=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (JHiltbrand) [reply CustomOptimizer.py:15]: Good point, I will rename and use "other" over "default".

File: CustomOptimizer.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+               coefficients['beta_2_t'],
+               coefficients['epsilon'],
+               grad,
+               use_locking=self._use_locking)
+
+    # This overridden method makes a dictionary "apply_state" that maps
+    # a model variable tuple to a dictionary that will contain
+    # information related to the learning rate-related hyperparmeters.
+    # E.g. each layer (kernel/bias/etc) specified in the model will appear in the var list
+    # therefore, the dictionary maps the unique layer name, GPU device, and type to
```
[the remark this answers is not in the document]
Comment (JHiltbrand) [reply CustomOptimizer.py:78]: Tpyo has been fixed.

File: DataLoader.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        batchMassReg  = np.empty([1,1]) 
+        batchModel    = np.empty([1,1])
+
+        initialized = False
+        for mnjet, data in self.data.items():
+            
+            batchIndices = None; factor = -1
+            if "BKG" in mnjet or ("EVTS" in mnjet and "_0" in mnjet):
+                factor = self.bkgSampleFactor
+                #print(mnjet, self.bkgSampleFactor)
```
[the remark this answers is not in the document]
Comment (JHiltbrand) [reply DataLoader.py:167]: Clean up on aisle 5.

File: train.py (hunk 1 above)
Comment (JHiltbrand) [reply train.py:236]: Yes sir

File: CustomLayer.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import tensorflow.keras as K
+from keras.engine.topology import Layer
+
+class FreeParamLayer(Layer):
+
+    def __init__(self, **kwargs):
+        super(FreeParamLayer, self).__init__(**kwargs)
```
[the remark this answers is not in the document]
Comment (JHiltbrand) [reply CustomLayer.py:7]: I'll believe you 😛

File: CustomLayer.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+    def build(self, input_shape):
+        self._freeParam = self.add_weight(name="freeParam",
+                                  shape=(1,),
+                                  initializer = K.initializers.RandomUniform(minval=0.0, maxval=1.0),
+                                  trainable=True)
+
+        super(FreeParamLayer, self).build(input_shape)
```
[the remark this answers is not in the document]
Comment (JHiltbrand) [reply CustomLayer.py:15]: Aye, aye, captain

**Review by cmadrid1 — APPROVED** [review]

**Review by cros0400 — APPROVED** [review]
